add a crude CLI that mimics CRMC

HDembinski opened the pull request:
First step towards a CRMC-like cli. This is a rather crude version and does not have full compatibility with crmc. However, the same arguments are accepted and it should be usable to generate hepmc and root files.

HDembinski commented:
@jncots @afedynitch If you want to try this out: just install impy as you normally would. This now makes a command `impy` available that supports a CRMC-like interface. Try `impy -h`.

HDembinski commented:
@jncots I just recently said in a review that we should not add a dependency just to draw nice progress bars for the file download. I haven't changed my stance on this, but everything is a trade-off, and so I now added `rich` to make the cli pretty, which also provides pretty progressbars. I use the progressbar in the cli to indicate progress, but also enabled it during file download.

Since `rich` provides more functionality than just a progressbar and since it is incredibly popular and well designed, I think it is safe to depend on `rich`.

HDembinski commented:
I fixed #86 as a byproduct of testing the cli. The implementation of `setunstable` in the EPOS fortran interface was incorrect, while `setstable` was correct. I remove both the Fortran implementations of `setstable` and `setunstable` and replaced them with Python implementations, because it is easy to handle this in Python and it is not performance-critical.

We should prefer interface code written in Python where performance considerations allow it, because it is easier to debug and fix Python code compared to Fortran code.

jncots commented on tests/test_cli.py:
```diff
@@ -0,0 +1,264 @@
+        format,
+        stdout=f"Format[ \t]*{format}",
+        file=f"impy_eposlhc_1_2212_2212_100.{ext}",
+    )
```
I get an error " undefined symbol: Npy__PyLong_AsInt" for most of the tests. What could be  the reason here?

HDembinski replied:
When and where? I did not see these and they are not listed in the CI runs. Perhaps your build is corrupted and you need to rebuild from scratch. This undefined symbol indicates that libpython is not linked properly into impy on your machine. This cannot have anything to do with this PR, it is a build system issue.

jncots replied:
This was a problem with pyf file from Epos. After regeneration of the pyf file tests pass.

HDembinski replied:
So it can be resolved?

jncots commented on tests/test_util.py:
```diff
@@ -0,0 +1,24 @@
+    assert util.tolerant_string_match("123", "abc1 s2.d34")
+    assert not util.tolerant_string_match("123", "321")
+    assert not util.tolerant_string_match("123", "3123")
+    assert not util.tolerant_string_match("123", "124")
```
Oh, now I get it what `tolerant_string_match` does.

jncots reviewed (commented):
Hello, @HDembinski. CLI is very cool feature. Indeed, `rich` library makes output look professional. The only thing is that the produced warnings obstructs the output. I also got errors `undefined symbol: Npy__PyLong_AsInt` during `cli` testing.

HDembinski commented:
> I also got errors `undefined symbol: Npy__PyLong_AsInt` during `cli` testing.

When and where? I didn't see these and the CI tests pass, too. Perhaps your build is corrupted and you need to rebuild from scratch.

HDembinski commented on src/fortran/EPOS-LHC/sources/epos-bas-lhc.f:
```diff
@@ -2831,7 +2831,7 @@ subroutine ainit
 
       if(inicnt.eq.1)then
-        write(ifmt,'(a)')'initializations ...'
+c        write(ifmt,'(a)')'initializations ...'
```
I think it is fine to make trivial changes like this, since this is not compromising the results of EPOS-LHC in any way. I think Anatoli also made changes to models by adding impy logging functions.

HDembinski commented on src/impy/__init__.py:
```diff
@@ -1,18 +1,19 @@
+from . import models  # noqa
+from . import kinematics  # noqa
+from . import constants  # noqa
 
```
I prefer relative imports because the syntax makes it immediately clear that we are importing from our own package and not from third-party. It also avoids the redundant `from impy ...`.

Pep8 recommends absolute imports, but also admits relative imports.  
https://peps.python.org/pep-0008/

HDembinski replied:
This is now resolved, absolute imports are used everywhere.

HDembinski commented on src/impy/cli.py:
```diff
@@ -0,0 +1,348 @@
+        if v is M:
+            VALID_MODELS.append(f"{M.label} [{k}]")
+            break
+    else:
```
No, this is correct. The else branch is triggered when the break inside the for loop is not triggered. That's what I want here.

HDembinski commented on src/impy/cli.py:
```diff
@@ -0,0 +1,348 @@
+    # "lhegz",
+)
+VALID_FORMATS = f"{{{', '.join(FORMATS)}}}"
+
```
The inner brace is for the f-string. The double outer braces are an escaped brace.

HDembinski commented on src/impy/cli.py:
```diff
@@ -0,0 +1,348 @@
+    return _FromParticleName._get_pdg(x)
+
+
+def particle_name(pid):
```
Right, it is a leftover that I will remove.

HDembinski commented on src/impy/cli.py:
```diff
@@ -0,0 +1,348 @@
+            if x == y:
+                Model = M
+                break
+            if tolerant_string_match(x, y):
```
I added some comments and you can also look into the unit test to see some examples `test_model...` in `test_cli.py`.

For several good reasons, we should specify models not by number (not human-friendly) but by name. We could just ask the user to pass the exact model label as a string (what Model.label returns), but that's not very user friendly. This code here
- Ignores case, so one does not have to remember typing SIBYLL or Sibyll.
- Allows you to write just "epos" to match "EPOS-LHC", because that's the only version of EPOS we have. This is done by requiring that all characters in the spec are in the model label and in correct order. See the unit test for `tolerant_string_match` in `test_util.py` for a few examples.
- The spec can be ambiguous, for example, when you request `sibyll`, in that case the code tells you what matches are found, this allows you to refine your spec.
- As a special case, if your spec matches a model exactly, you will get that model even if there is another model that would match `tolerant_string_match`, too. Without that rule, you could not select "Sibyll-2.3" because it is a substring of "Sibyll-2.3d"

HDembinski commented on src/impy/cli.py:
```diff
@@ -0,0 +1,348 @@
+                if task_id is None:
+                    task_id = bar.add_task("", total=args.number)
+                bar.advance(task_id, 1)
+
```
Not anymore, please upgrade pyhepmc.

HDembinski replied:
For now, our only option is to turn messages off that are produced by the Fortran code during event generation. `rich` in principle allows one to write info messages without messing up the progressbar, but that only works if the Python print function is used. The Fortran code uses a different IO system that `rich` has no control over.

HDembinski commented:
@jncots @afedynitch This PR is ready to merge. Could one of you please approve it?

afedynitch reviewed (commented):
Very cool PR.

I have two problems that I can not investigate by myself right now (on a trip):
1) When trying to run SIBYLL through the CLI on my M1 mac, I get (guess it's related to the cross section readout)
```
anatoli@Anatolis-MacBook-Pro impy % impy -m sibyll-2.1 -n 10000 -S 1000 -o root -f hey.root
ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
```
2) I would have run the tests, but pyhepmc doesn't build 3.11 wheels (and my mac needs 3.11 python for some specific stuff right now, and I really don't want to mess up the running system).  And pyhepmc doesn't build from srcdist due to a very weird error:
```Collecting pyhepmc
  Using cached pyhepmc-2.9.0.tar.gz (356 kB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Installing backend dependencies ... done
  Preparing metadata (pyproject.toml) ... done
Discarding https://files.pythonhosted.org/packages/6b/7e/ee0266b1b6557fac37796e82a1f1bca9d42ee453254a7b403a2e6370f54e/pyhepmc-2.9.0.tar.gz (from https://pypi.org/simple/pyhepmc/) (requires-python:>=3.7): Requested pyhepmc from https://files.pythonhosted.org/packages/6b/7e/ee0266b1b6557fac37796e82a1f1bca9d42ee453254a7b403a2e6370f54e/pyhepmc-2.9.0.tar.gz has inconsistent version: expected '2.9.0', but metadata has '0.0.0'
```

Sorry for posting this here although it's more about the pyhepmc repo.

Apart from these I'm very happy that you've built the CLI.

One change to logic is a bit strange: return True for success is ok but a bit uncommon (in many fortran codes people use 0 for all fine or !=0 for not ok + statuscode). Also in C or *NIX int main() returns 0 for everything ok and otherwise a non-zero return code. Technically PHOJET can return more than a non-zero rejection flag., indicating what went wrong. There is way more info needed to describe if something goes wrong rather than in the case where things are simply "ok". Also, checking for error by `if not _generate_event` sounds more logical to me. I don't think that turning the returns into `True`'s in `generate_event` has any foreseeable issue - so feel free to decide.

HDembinski commented on src/impy/cli.py:
```diff
@@ -0,0 +1,356 @@
+# Impy-exclusive models do not get a number and should not be added here.
+MODELS = {
+    0: models.EposLHC,
+    # 1: models.Epos199,
```
Yes, this seems to correspond to what I saw in the EPOS-LHC code, you can set the tune at the beginning. Please add an issue if you want to support EPOS-1.99.

afedynitch replied:
ok, thanks

HDembinski commented:
> Very cool PR.
> 
> I have two problems that I can not investigate by myself right now (on a trip):
> 
>     1. When trying to run SIBYLL through the CLI on my M1 mac, I get (guess it's related to the cross section readout)
> 
> 
> ```
> anatoli@Anatolis-MacBook-Pro impy % impy -m sibyll-2.1 -n 10000 -S 1000 -o root -f hey.root
> ValueError: The truth value of an array with more than one element is ambiguous. Use a.any() or a.all()
> ```

Good catch, that was a bug in the new cross_section code. This is now fixed and a proper test was added. I removed the broken support for computing the cross-section for nuclear targets for now. The code that was implemented does not work. This needs further investigation.

>     2. I would have run the tests, but pyhepmc doesn't build 3.11 wheels (and my mac needs 3.11 python for some specific stuff right now, and I really don't want to mess up the running system).  And pyhepmc doesn't build from srcdist due to a very weird error:
> 
> 
> ```
>   Using cached pyhepmc-2.9.0.tar.gz (356 kB)
>   Installing build dependencies ... done
>   Getting requirements to build wheel ... done
>   Installing backend dependencies ... done
>   Preparing metadata (pyproject.toml) ... done
> Discarding https://files.pythonhosted.org/packages/6b/7e/ee0266b1b6557fac37796e82a1f1bca9d42ee453254a7b403a2e6370f54e/pyhepmc-2.9.0.tar.gz (from https://pypi.org/simple/pyhepmc/) (requires-python:>=3.7): Requested pyhepmc from https://files.pythonhosted.org/packages/6b/7e/ee0266b1b6557fac37796e82a1f1bca9d42ee453254a7b403a2e6370f54e/pyhepmc-2.9.0.tar.gz has inconsistent version: expected '2.9.0', but metadata has '0.0.0'
> ```
> 
> Sorry for posting this here although it's more about the pyhepmc repo.

Yes, that's strange. I installed python-3.11 with brew and `pip install pyhepmc` works. I also have a M1 mac. `pip list` shows the version of pyhepmc correctly. Something to look into when we start to support 3.11 in CI, currently we do not. That is a topic for another PR.

> Apart from these I'm very happy that you've built the CLI.
> 
> One change to logic is a bit strange: return True for success is ok but a bit uncommon (in many fortran codes people use 0 for all fine or !=0 for not ok + statuscode). Also in C or *NIX int main() returns 0 for everything ok and otherwise a non-zero return code. Technically PHOJET can return more than a non-zero rejection flag., indicating what went wrong. There is way more info needed to describe if something goes wrong rather than in the case where things are simply "ok". Also, checking for error by `if not _generate_event` sounds more logical to me. I don't think that turning the returns into `True`'s in `generate_event` has any foreseeable issue - so feel free to decide.

That's a fair point, I will revert my change.

afedynitch commented:
OK, thanks, good to go. Feel free to merge.

P.S.: I could get pyhepmc installed by passing [`--use-deprecated=legacy-resolver`](https://stackoverflow.com/questions/67074684/pip-has-problems-with-metadata) to pip on 3.11. The rest seems to work under 3.11.
